src/hotspot/os/linux/os_linux.cpp

```diff
@@ -911,6 +915,29 @@ bool os::create_thread(Thread* thread, ThreadType thr_type,
   // Calculate stack size if it's not specified by caller.
   size_t stack_size = os::Posix::get_initial_stack_size(thr_type, req_stack_size);
   size_t guard_size = os::Linux::default_guard_size(thr_type);
+
+  // If THPs are unconditionally enabled, the following scenario can lead to huge RSS:
+  // - parent thread spawns, in quick succession, multiple child threads
+  // - child threads are slow to start
+  // - thread stacks of future child threads are adjacent and get merged into one large VMA
+  //   by the kernel, and subsequently transformed into huge pages by khugepaged
+  // - child threads come up, place JVM guard pages, thus splinter the large VMA, splinter
+  //   the huge pages into many (still paged-in) small pages.
+  // The result of that sequence are thread stacks that are fully paged-in even though the
+  // threads did not even start yet.
+  // We prevent that by letting the glibc allocate a guard page, which causes a VMA with different
+  // permission bits to separate two ajacent thread stacks and therefore prevent merging stacks
+  // into one VMA.
+  if (PreventTHPsForThreadStacks) {
+    guard_size = MAX2(guard_size, os::vm_page_size());
+    // Add an additional page to the stack size to reduce its chances of getting huge page aligned
+    // so that the stack does not get backed by a transparent huge page.
```

Comment:
I don't think these two adjustments should be combined like this. The existing +1 may be sufficient for some folk who do not want the cost of creating a glibc guard page as well.

Reply:
I disagree, see earlier remarks. "May be sufficient" is vague - either you want to prevent THPs forming in thread stacks, or you are okay with it. There is no point in doing one without the other.

If you want to prevent them, you have to do both since the effects laid out in this PR are not predictable. Even if your app does not create - and why would you care for that one guard page then - the JVM creates a bunch of threads in quick succession too and thus suffers from the same effects.

And if you want to have THPs, then you should disable both mitigations together.